### ISSUE TITLE
Landscape - Top menu left over rendering

### ISSUE THREAD

**Issue by mergen3107** [koreader/koreader#11303]
* KOReader version: v2023.10-55
* Device: Kindle Scribe (5.16.2)

#### Issue
In landscape mode, top menu overlaps when switching tabs.

#### Steps to reproduce
Go to landscape with a gesture - using a profile in Quick Menu to rotate by 90 deg CW/CCW or to absolute 90/270 deg.

Open top menu and skim through menu tabs.

The tabs with more items will stay when going to tab with fewer items. 

However, the issue is only in rendering, it seems - tapping on this leftover menu is normal (close menu).

See photo:
![image](https://github.com/koreader/koreader/assets/6279855/0fa6126e-ae6a-4fa8-958e-06658a1b6494)

##### `crash.log` (if applicable)
On mobile right now, will provide logs later if needed.

**Comment by NiLuJe** [comment]
Is the rotation method relevant, or does it also happen with a bottom menu rotation?

**Comment by NiLuJe** [comment]
No matter how I rotate, I can't repro an a device that happens to be testing #11297, which has a very vague chance of being relevant here?

**Comment by NiLuJe** [comment]
So, yeah, going to need verbose debug logs, and potentially added logging around this line:

https://github.com/koreader/koreader/blob/5d2a4410648dfc993e6dbc01cbf23941a92404c3/frontend/ui/widget/touchmenu.lua#L759

specifically, dumping sboth `self.dimen` and `old_dimen`.

I would also attempt to rule out the startup rotation shenanigans, just in case...

**Comment by mergen3107** [comment]
> Is the rotation method relevant, or does it also happen with a bottom menu rotation?

1) Bottom menu rotation is OK.
2) I realized, +/-90 deg rotations are OK too.

Only absolute value rotations are not OK.

**Comment by mergen3107** [comment]
> on a device that happens to be testing #11297

Let me get logs first, then I can that out :D

**Comment by mergen3107** [comment]
[rotation_landsacpe_topMenu.txt](https://github.com/koreader/koreader/files/13797278/rotation_landsacpe_topMenu.txt)

Look for `LOGROT` for added debug lines

**Comment by mergen3107** [comment]
OK, more details. Only _one_ rotation is bad - when the the wide bezel is on the bottom.

When looking the at device at `U`, this side is on the left, so device needs to be rotated CCW to get there.

**Comment by mergen3107** [comment]
Further testing, I unlocked gyro and rotated to that position - same issue.

**Comment by mergen3107** [comment]
It is not KOL vs KUAL, and I tried to move `Kindle:init()` back to the bottom - no difference.

**Comment by mergen3107** [comment]
[rotation_landsacpe_topMenu_with_11297_diff.txt](https://github.com/koreader/koreader/files/13797312/rotation_landsacpe_topMenu_with_11297_diff.txt)

This is with 11297 applied as a diff.

Open KOReader, rotate to that orientation with bezel on the bottom, open top menu, go through tabs one by one.

Interestingly, every second tab "freezes": eink icon fills up as if it is "pressed", but tab doesn't switch. Hitting again opens the tab. Also, hamburger menu - Exit button freezes as well.

So in the end I exited menu, rotated back to `U` and exited via top menu.

**Comment by mergen3107** [comment]
I also tried reverting my `swipe_direction` patch, no difference.

It seems to me this issue might have been there for a long time, I just rarely skimmed through the top menu in landscape :D

**Comment by mergen3107** [comment]
> > Is the rotation method relevant, or does it also happen with a bottom menu rotation?
> 
> 1. Bottom menu rotation is OK.
> 2. I realized, +/-90 deg rotations are OK too.
> 
> Only absolute value rotations are not OK.

**UPDATE:** OK, more details. Only one rotation is bad - when the the wide bezel is on the bottom.

When looking the at device at U, this side is on the left, so device needs to be rotated CCW to get there.

**Comment by mergen3107** [comment]
It is the same story in FM: only `.DEVICE_ROTATED_COUNTER_CLOCKWISE` is bad.

**Comment by mergen3107** [comment]
Bingo! The culprit is enabled `HW dithering` O_O

**Comment by mergen3107** [comment]
Only if `Disable SW dithering` is checked (HW dithering on) - top menu issues are present,

If either `Disable HW dithering` is check or both, no issues.

**Comment by NiLuJe** [comment]
Huh. Yeah, every non-full screen dithered refresh ioctl fails with EBADFD.

**Comment by mergen3107** [comment]
I looked around for `hw_dithering` in files, and only found this place:

https://github.com/koreader/koreader/blob/5d2a4410648dfc993e6dbc01cbf23941a92404c3/frontend/ui/uimanager.lua#L1291-L1300

**Comment by NiLuJe** [comment]
Mildly confused why you're getting dithered refresh at all, though. Is there an image on the page you're on?

**Comment by mergen3107** [comment]
> Mildly confused why you're getting dithered refresh at all, though. Is there an image on the page you're on?

oh, yes! I had a PDF doc open with image in it, then tried in a fb2 book (with image as well), and in FM there are thumbnails of covers in detailed view.

**Comment by mergen3107** [comment]
However, it is weird that it only fails in one orientation out of four, and it landscape (not `D` with most problems or something)

**Comment by NiLuJe** [comment]
There might be weirder alignment constraints than on mxcfb, or... something.

You can maybe check if bumping the alingment constraints to 16 or 32 or 64 instead of 8 help?

https://github.com/koreader/koreader-base/blob/d83148b27e25b9332678221f6737d66be7e73318/ffi/framebuffer_mxcfb.lua#L329-L343

**Comment by NiLuJe** [comment]
Or it's just broken, period. Remember, it doesn't work at all on Bellatrix ;).

**Comment by mergen3107** [comment]
> You can maybe check if bumping the alingment constraints to 16 or 32 or 64 instead of 8 help?

Should I just print `x`, `x_orig`, etc in good and bad orientations?

**Comment by NiLuJe** [comment]
> > You can maybe check if bumping the alingment constraints to 16 or 32 or 64 instead of 8 help?
> 
> Should I just print `x`, `x_orig`, etc in good and bad orientations?

No, just bump the `8` constant in the four ALIGN_ calls.

**Comment by NiLuJe** [comment]
I also don't recall if the stock kernels log, but I think they do, so, check `dmesg` after a failed ioctl.

**Comment by mergen3107** [comment]
> No, just bump the `8` constant in the four ALIGN_ calls.

Nope, no dice. Tried 16, 32, 64.


> I also don't recall if the stock kernels log, but I think they do, so, check `dmesg` after a failed ioctl.

```
[ 2485.795429] -(0)[8471:reader.lua]Function entered at [<c0299c60>] from [<c029a5d4>]
[ 2485.795432] -(0)[8471:reader.lua]Function entered at [<c029a5d4>] from [<c0108a40>]
[ 2485.795439] .(0)[8471:reader.lua][HWTCON ERR]invalid rectangle:[0 -60 2480 960] @hwtcon_rect_check_valid,42
[ 2485.795444] .(0)[8471:reader.lua][HWTCON ERR]invalid region[0 -60 2480 960] panel[2480 1860] rot:3 @hwtcon_fb_check_update_data_invalid,274
[ 2486.359493] .(1)[8471:reader.lua]bq256xx-charger 0-006b: bq25611_get_charger_status, return 2, status0:status1:status2=0x0:0x0:0x0
```

**Comment by NiLuJe** [comment]
Oh, yeah, that logic needs to be smarter.

`992x2480 region @ (872, 0)`

872 + 992 = 1864. That's larger than the panel size, so the driver says "nope".

I'll need a working brain for that; in the mean time, try dropping that whole `if dither then ALIGN stuff` block, it's likely unnecessary on those SoCs.

**Comment by mergen3107** [comment]
`[HWTCON ERR]invalid region[0 -60 2480 960] panel[2480 1860] rot:3 @hwtcon_fb_check_update_data_invalid,274`
What? how can the `y` coordinate be negative?

**Comment by mergen3107** [comment]
> In the mean time, try dropping that whole `if dither then ALIGN stuff` block, it's likely unnecessary on those SoCs.

That did it! No more repaint issues in the top menu

**Comment by mergen3107** [comment]
Thank you @NiLuJe !

**Comment by NiLuJe** [comment]
Untested as of yet, but updated code pushed @ https://github.com/NiLuJe/koreader-base

**Comment by NiLuJe** [comment]
Well, technically, barely tested by throwing random stuff at this: 

```lua
#!/bin/env luajit

local function getWidth()
    return 1860
end

local function getHeight()
    return 2480
end

local band = bit.band
local bnot = bit.bnot
local floor = math.floor
local ceil = math.ceil

-- A couple helper functions to compute aligned values...
-- c.f., <linux/kernel.h> & ffi/framebuffer_linux.lua
local function ALIGN_DOWN(x, a)
    -- x & ~(a-1)
    local mask = a - 1
    return band(x, bnot(mask))
end

local function ALIGN_UP(x, a)
    -- (x + (a-1)) & ~(a-1)
    local mask = a - 1
    return band(x + mask, bnot(mask))
end

--[[
More specific checkBounds variant that will return the specified rect bounded inside the blitbuffer.
i.e., make sure that x & y don't go OOB on either side, and that x+w & y+h don't go over their respective dimensions.

@param x coordinate
@param y coordinate
@param w dimension
@param h dimension
@param alignment *optional* alignment constraint, if not nil, *must* be > 0 and a power of two!

@return rect strictly bounded inside the bb
--]]
local function getBoundedRect(x, y, w, h, alignment)
    local max_w = getWidth()
    local max_h = getHeight()

    -- Deal with OOB coordinates
    if x >= max_w then
        x = 0
        w = 0
    end
    if y >= max_h then
        y = 0
        h = 0
    end

    -- Deal with negative coordinates
    if x < 0 then
        w = w + x
        x = 0
    end
    if y < 0 then
        h = h + y
        y = 0
    end

    -- Align to pixel grid
    x = floor(x)
    y = floor(y)
    w = ceil(w)
    h = ceil(h)

    -- Honor alignment constraints, if any.
    -- coordinates can only go down, but never below 0, so there's no risk of it invalidating our previous OOB checks.
    -- NOTE: c.f., dithering comments in framebuffer_mxcfb for a potential use-case
    if alignment then
        local x_orig = x
        x = ALIGN_DOWN(x_orig, alignment)
        local x_fixup = x_orig - x
        w = ALIGN_UP(w + x_fixup, alignment)
        local y_orig = y
        y = ALIGN_DOWN(y_orig, alignment)
        local y_fixup = y_orig - y
        h = ALIGN_UP(h + y_fixup, alignment)
    end

    -- Make sure the rect fits strictly inside the bb
    if x + w > max_w then
        w = max_w - x
    end
    if y + h > max_h then
        h = max_h - y
    end

    return x, y, w, h
end

local x = tonumber(arg[1]) or 0
local y = tonumber(arg[2]) or 0
local w = tonumber(arg[3]) or getWidth()
local h = tonumber(arg[4]) or getHeight()
local a = tonumber(arg[5]) or nil

local bx, by, bw, bh = getBoundedRect(x, y, w, h, a)
print(string.format("%dx%d+%d+%d => %dx%d+%d+%d", w, h, x, y, bw, bh, bx, by))
```

**Comment by mergen3107** [comment]
I'll trst it, thanks :D

**Comment by mergen3107** [comment]
@NiLuJe 
I updated to pure v2023.10-82 and applied diff from your new 4 commits.
Patch was successful.

Here what I got:
```
./luajit: ./ffi/blitbuffer.lua:1578: 'struct BlitBuffer8' has no member named 'getBoundedRect'
stack traceback:
	./ffi/blitbuffer.lua:1578: in function 'paintRect'
	./ffi/blitbuffer.lua:1862: in function 'paintBorder'
	./ffi/blitbuffer.lua:1909: in function 'paintRoundedRect'
	frontend/ui/widget/container/framecontainer.lua:111: in function 'paintTo'
	frontend/ui/widget/container/movablecontainer.lua:211: in function 'paintTo'
	frontend/ui/widget/container/centercontainer.lua:31: in function 'paintTo'
	frontend/ui/widget/container/inputcontainer.lua:83: in function 'paintTo'
	frontend/ui/widget/infomessage.lua:281: in function 'paintTo'
	frontend/ui/uimanager.lua:1257: in function '_repaint'
	frontend/ui/uimanager.lua:1322: in function 'forceRePaint'
	frontend/apps/reader/readerui.lua:616: in function 'showReaderCoroutine'
	frontend/apps/reader/readerui.lua:605: in function 'showReader'
	./reader.lua:264: in main chunk
	[C]: at 0x00013ebd
lipc-wait-event exited normally with status: 0
```

**Comment by mergen3107** [comment]
https://github.com/koreader/koreader-base/compare/master...NiLuJe:koreader-base:master#diff-4c53846f9e82fcc8f6ecf4c8c885b46c21a98f0cc3121a067ef2ffb853f221e2R1070

I changed
`function BB_mt:getBoundedRect(x, y, w, h, alignment)` to
`function BB_mt.__index:getBoundedRect(x, y, w, h, alignment)`
and it worked! 

I can confirm that all orientation work great now. Thank you!
Happy New Year!

**Comment by mergen3107** [comment]
@NiLuJe 
I pressed power button to put Kindle to sleep in KOReader, and crash happened:
```
./luajit: ./ffi/blitbuffer.lua:1075: attempt to compare number with nil
stack traceback:
	./ffi/blitbuffer.lua:1075: in function 'getBoundedRect'
	./ffi/framebuffer_mxcfb.lua:264: in function 'mech_refresh'
	./ffi/framebuffer_mxcfb.lua:724: in function 'refreshFull'
	frontend/ui/screensaver.lua:532: in function 'show'
	frontend/device/kindle/device.lua:289: in function 'intoScreenSaver'
	frontend/device/kindle/device.lua:389: in function 'handler'
	frontend/ui/uimanager.lua:1434: in function 'handleInputEvent'
	frontend/ui/uimanager.lua:1534: in function 'handleInput'
	frontend/ui/uimanager.lua:1578: in function 'run'
	./reader.lua:265: in main chunk
	[C]: at 0x00013ebd
lipc-wait-event exited normally with status: 0

```

**Comment by mergen3107** [comment]
Added some `logger.dbg`:
```
12/30/23-18:28:56 DEBUG LOGG max_w = 1860 LOGG max_h = 2480 LOGG x = nil LOGG y = nil 
```
So, the screensaver passes on `nil` for both `x` and `y`.

**Comment by mergen3107** [comment]
I pressed power button, then light went out, but I still saw the book text, then after 1-2 second the screenshot appeared.

But when I press the power button again, I am greeted with Kindle home screen.

**Comment by mergen3107** [comment]
Here is a full log:
[dither.crash.txt](https://github.com/koreader/koreader/files/13799412/dither.crash.txt)

**Comment by mergen3107** [comment]
The end looks like this:
```
12/30/23-18:28:56 DEBUG AutoSuspend: onInputEvent 
12/30/23-18:28:56 DEBUG refresh: full nil nil nil nil nil 
12/30/23-18:28:56 DEBUG LOGG max_w = 1860 LOGG max_h = 2480 LOGG x = nil LOGG y = nil
```
So it crashes after I hit power button the first time.

**Comment by mergen3107** [comment]
Is this because this call does not pass any arguments?
https://github.com/koreader/koreader/blob/5d2a4410648dfc993e6dbc01cbf23941a92404c3/frontend/ui/screensaver.lua#L532

It refers to this function
https://github.com/koreader/koreader-base/blob/d83148b27e25b9332678221f6737d66be7e73318/ffi/framebuffer.lua#L170-L174
which is then overridden by
https://github.com/koreader/koreader-base/blob/d83148b27e25b9332678221f6737d66be7e73318/ffi/framebuffer_mxcfb.lua#L772-L775
and it expects `(x,y,w,h,d)` as input.

Log shows those were `nil`.

**Comment by mergen3107** [comment]
Other than that I found two other places where `refreshFull()` has no input arguments:
https://github.com/koreader/koreader/blob/5d2a4410648dfc993e6dbc01cbf23941a92404c3/frontend/device/generic/device.lua#L329
https://github.com/koreader/koreader/blob/5d2a4410648dfc993e6dbc01cbf23941a92404c3/frontend/device/generic/device.lua#L374

**Comment by NiLuJe** [comment]
Yup, nice catch, fixed that one in a couple of places.

I dropped the nil guards entirely, because UIManager will never call this without a rect, so it made no sense to be defensive.

TL;DR: `Screen:refreshFull()` =>`Screen:refreshFull(0, 0, Screen:getWidth(), Screen:getHeight())`

**Comment by mergen3107** [comment]
How about the fifth inout parameter, `d` which is the dithering? In the log above it was also nil, however I am pretty sure a screensaver image is an image :D

**Comment by mergen3107** [comment]
So that `hw_dither` is supposed to kick in, right?

**Comment by mergen3107** [comment]
Thanks @NiLuJe,

With your 9 commits + `Screen:refreshFull(0, 0, Screen:getWidth(), Screen:getHeight())` in the `screensaver.lua`, screensaver works!

Should I also replace those calls in `generic/device.lua` as well?
(https://github.com/koreader/koreader/issues/11303#issuecomment-1872636810)

**Comment by mergen3107** [comment]
It says in `generic/device.lua` in that function, `-- Only used on platforms where we handle suspend ourselves.`. Does this apply to Kindles?

**Comment by NiLuJe** [comment]
https://github.com/koreader/koreader-base/pull/1718 + https://github.com/koreader/koreader/pull/11307

Flagged for after the release, because that touches a fairly ubiquitous codepath, so it could have unforeseen circumstances...

(Unless we're not due a release for a while, @Frenzie?).

**Comment by mergen3107** [comment]
> Didn't see anything obviously awful after a quick test on a Clara 2E & a Forma (which is the only device I know _needs_ the alignment fix).

I continue testing with Kindle Scribe. So far, so good with all new changes.

Anything else super-specific to framebuffer I need to test? :D

**Comment by NiLuJe** [comment]
The change technically touches *all* the code we use to compute rectangles (when drawing, blitting or refreshing stuff based on rectangles), so, it could in theory affect a lot of things, possibly fairly hard to notice (like +/- 1px positioning differences for callers that still pass subpixel coordinates, as I tweaked the rounding (floor coords, ceil dims; like Geom itself). We fixed *a lot* of these, though, so it shouldn't really change anything in practice).

**Comment by mergen3107** [comment]
As a side note, when I leave only SW dithering on (`Disable HW dither` checked, `Disable SW dither` unchecked), it seems dithering doesn't work at all anymore in mupdf - toggling bottom menu Dithering options does nothing.
The same effect is when both `Disable HW dither` and `Disable SW dither` are checked.

When I don't have `canHWDither = yes` in the Kindle definition, SW dithering works. Is this by design?

**Comment by mergen3107** [comment]
> We fixed _a lot_ of these, though, so it shouldn't really change anything in practice

Alright, I'll keep testing :D 
Thank you very much for your kind help!

**Comment by NiLuJe** [comment]
> When I don't have `canHWDither = yes` in the Kindle definition, SW dithering works. Is this by design?

Yes, when the device supports HW dithering, the button will (attempt to) use it, regardless of the dev override.

(The label doesn't change, but you cna long-press it, and the help text will detail whether it's hardware or software).

**Comment by mergen3107** [comment]
> (The label doesn't change, but you can long-press it, and the help text will detail whether it's hardware or software).

Ah, I see.
So it turns out, if HW dithering is supported by the device, only HW dithering can be used in mupdf. The bottom menu does not change when I leave only SW dithering on.

**Comment by mergen3107** [comment]
> > (The label doesn't change, but you can long-press it, and the help text will detail whether it's hardware or software).
> 
> Ah, I see. So it turns out, if HW dithering is supported by the device, only HW dithering can be used in mupdf. The bottom menu does not change when I leave only SW dithering on - it still says "Enable hardware dithering".

**Comment by NiLuJe** [comment]
Yup, exactly.
